File: formcraft-plm/src/main/java/fr/formcraft/model/enums/ProductState.java
package fr.formcraft.model.enums;

/**
 * Product lifecycle state machine states.
 * Valid transitions: DRAFT -> IN_VALIDATION -> VALIDATED -> ARCHIVED
 */
public enum ProductState {
    DRAFT,
    IN_VALIDATION,
    VALIDATED,
    ARCHIVED;

    public boolean canTransitionTo(ProductState target) {
        return switch (this) {
            case DRAFT         -> target == IN_VALIDATION;
            case IN_VALIDATION -> target == VALIDATED || target == DRAFT;
            case VALIDATED     -> target == ARCHIVED;
            case ARCHIVED      -> false;
        };
    }
}

File: app/enums/product_state.py
import enum


class ProductState(str, enum.Enum):
    DRAFT = "DRAFT"
    IN_VALIDATION = "IN_VALIDATION"
    VALIDATED = "VALIDATED"
    ARCHIVED = "ARCHIVED"

    def can_transition_to(self, target: "ProductState") -> bool:
        match self:
            case ProductState.DRAFT:
                return target == ProductState.IN_VALIDATION
            case ProductState.IN_VALIDATION:
                return target in (ProductState.VALIDATED, ProductState.DRAFT)
            case ProductState.VALIDATED:
                return target == ProductState.ARCHIVED
            case ProductState.ARCHIVED:
                return False
